pkg/controller/machineset/machineset_controller.go

```diff
@@ -207,48 +395,331 @@ func (c *MachineSetController) handleErr(err error, key interface{}) {
+	logger := c.logger.WithField("machineset", key)
+	logger.Debugln("Started syncing machine set")
 	defer func() {
-		glog.V(4).Infof("Finished syncing node group %q (%v)", key, time.Since(startTime))
+		logger.WithField("duration", time.Since(startTime)).Debugln("Finished syncing machine set")
 	}()
 
 	ns, name, err := cache.SplitMetaNamespaceKey(key)
 	if err != nil {
-		return err
+		utilruntime.HandleError(fmt.Errorf("cannot parse machine set key %q: %v", key, err))
+		return nil
 	}
 	if len(ns) == 0 || len(name) == 0 {
-		return fmt.Errorf("invalid machineset key %q: either namespace or name is missing", key)
+		utilruntime.HandleError(fmt.Errorf("invalid machineset key %q: either namespace or name is missing", key))
+		return nil
 	}
 
 	machineSet, err := c.machineSetsLister.MachineSets(ns).Get(name)
 	if errors.IsNotFound(err) {
-		glog.V(2).Infof("Node group %v has been deleted", key)
+		logger.Debugln("machine set has been deleted")
+		c.jobControl.ObserveOwnerDeletion(key)
 		return nil
 	}
 	if err != nil {
 		return err
 	}
 
-	// Deep-copy otherwise we are mutating our cache.
-	// TODO: Deep-copy only when needed.
-	ng := machineSet.DeepCopy()
+	shouldProvisionMachineSet := c.shouldProvision(machineSet)
+
+	jobFactory, err := c.getJobFactory(machineSet)
+	if err != nil {
+		return err
+	}
+
+	job, isJobNew, err := c.jobControl.ControlJobs(key, machineSet, shouldProvisionMachineSet, jobFactory, logger)
+	if err != nil {
+		return err
+	}
+
+	if !shouldProvisionMachineSet {
+		return nil
+	}
+
+	switch {
+	// New job has not been created, so an old job must exist. Set the cluster
+	// to not provisioning yet as the old job is deleted.
+	case job == nil:
+		return c.setMachineSetToNotProvisioning(machineSet)
+	// Job was not newly created, so sync machine set status with job.
+	case !isJobNew:
+		logger.Debugln("provisioning job exists, will sync with job")
+		return c.syncMachineSetStatusWithJob(machineSet, job)
+	// MachineSet should have a job to provision the current spec but it was not
+	// found.
+	case isMachineSetProvisioning(machineSet):
+		return c.setJobNotFoundStatus(machineSet)
+	// New job created for new provisioning
+	default:
+		return nil
+	}
+}
+
+func (c *MachineSetController) clusterForMachineSet(machineSet *clusteroperator.MachineSet) (*clusteroperator.Cluster, error) {
+	controllerRef := metav1.GetControllerOf(machineSet)
+	if controllerRef.Kind != clusterKind.Kind {
+		return nil, nil
+	}
+	cluster, err := c.clustersLister.Clusters(machineSet.Namespace).Get(controllerRef.Name)
+	if err != nil {
+		return nil, err
+	}
+	if cluster.UID != controllerRef.UID {
+		// The controller we found with this Name is not the same one that the
+		// ControllerRef points to.
+		return nil, nil
+	}
+	return cluster, nil
+}
+
+func (c *MachineSetController) shouldProvision(machineSet *clusteroperator.MachineSet) bool {
+	if machineSet.Status.ProvisionedJobGeneration == machineSet.Generation {
```

Comment:
Do we also need to add a check for whether the cluster spec has changed? Or do you expect that any change to the cluster spec that would change the deployment of the machine set would be accomplished by a separate controller (ClusterController?) deleting and re-creating the machineset object or making some other modification to the spec of the machineset?

Reply:
I would think (and prefer) the latter